File: instrumentation/okhttp3-websocket/library/src/main/java/io/opentelemetry/instrumentation/library/okhttp/v3_0/websocket/internal/WebsocketListenerWrapper.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.library.okhttp.v3_0.websocket.internal;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.opentelemetry.android.annotations.Incubating;
import io.opentelemetry.api.common.Attributes;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

@Incubating
public class WebsocketListenerWrapper extends WebSocketListener {
    private final WebSocketListener delegate;

    public WebsocketListenerWrapper(WebSocketListener delegate) {
        this.delegate = delegate;
    }

    @Override
    public void onClosed(@NonNull WebSocket webSocket, int code, @NonNull String reason) {
        Attributes attributes = WebsocketAttributeExtractor.extractAttributes(webSocket);
        WebsocketEventGenerator.generateEvent("websocket.close", attributes);
        delegate.onClosed(webSocket, code, reason);
    }

    @Override
    public void onOpen(@NonNull WebSocket webSocket, @NonNull Response response) {
        Attributes attributes = WebsocketAttributeExtractor.extractAttributes(webSocket);
        WebsocketEventGenerator.generateEvent("websocket.open", attributes);
        delegate.onOpen(webSocket, response);
    }

    @Override
    public void onMessage(@NonNull WebSocket webSocket, @NonNull String text) {
        Attributes attributes = WebsocketAttributeExtractor.extractAttributes(webSocket);
        WebsocketEventGenerator.generateEvent(
                "websocket.message",
                attributes.toBuilder()
                        .put(WebsocketAttributes.MESSAGE_TYPE, "text")
                        .put(WebsocketAttributes.MESSAGE_SIZE, text.length())
                        .build());
        delegate.onMessage(webSocket, text);
    }

    @Override
    public void onMessage(@NonNull WebSocket webSocket, @NonNull ByteString bytes) {
        Attributes attributes = WebsocketAttributeExtractor.extractAttributes(webSocket);
        WebsocketEventGenerator.generateEvent(
                "websocket.message",
                attributes.toBuilder()
                        .put(WebsocketAttributes.MESSAGE_TYPE, "bytes")
                        .put(WebsocketAttributes.MESSAGE_SIZE, bytes.size())
                        .build());
        delegate.onMessage(webSocket, bytes);
    }

    @Override
    public void onFailure(
            @NonNull WebSocket webSocket, @NonNull Throwable t, @Nullable Response response) {
        Attributes attributes = WebsocketAttributeExtractor.extractAttributes(webSocket);
        WebsocketEventGenerator.generateEvent("websocket.error", attributes);
        delegate.onFailure(webSocket, t, response);
    }
}

File: instrumentation/okhttp3-websocket/library/src/main/java/io/opentelemetry/instrumentation/library/okhttp/websocket/internal/WebsocketListenerWrapper.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.library.okhttp.websocket.internal

import io.opentelemetry.android.semconv.events.WebsocketCloseEvent
import io.opentelemetry.android.semconv.events.WebsocketErrorEvent
import io.opentelemetry.android.semconv.events.WebsocketMessageEvent
import io.opentelemetry.android.semconv.events.WebsocketOpenEvent
import io.opentelemetry.api.OpenTelemetry
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.logs.Logger
import io.opentelemetry.semconv.HttpAttributes
import io.opentelemetry.semconv.NetworkAttributes
import io.opentelemetry.semconv.UrlAttributes
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString

class WebsocketListenerWrapper(
    private val delegate: WebSocketListener,
) : WebSocketListener() {
    override fun onClosed(
        webSocket: WebSocket,
        code: Int,
        reason: String,
    ) {
        val attributes = extractAttributes(webSocket)
        WebsocketCloseEvent().emit(logger, attributes)
        delegate.onClosed(webSocket, code, reason)
    }

    override fun onOpen(
        webSocket: WebSocket,
        response: Response,
    ) {
        val attributes = extractAttributes(webSocket)
        WebsocketOpenEvent().emit(logger, attributes)
        delegate.onOpen(webSocket, response)
    }

    override fun onMessage(
        webSocket: WebSocket,
        text: String,
    ) {
        val attributes = extractAttributes(webSocket)
        WebsocketMessageEvent(
            websocketMessageSize = text.length.toLong(),
            websocketMessageType = "text",
        ).emit(logger, attributes)
        delegate.onMessage(webSocket, text)
    }

    override fun onMessage(
        webSocket: WebSocket,
        bytes: ByteString,
    ) {
        val attributes = extractAttributes(webSocket)
        WebsocketMessageEvent(
            websocketMessageSize = bytes.size.toLong(),
            websocketMessageType = "bytes",
        ).emit(logger, attributes)
        delegate.onMessage(webSocket, bytes)
    }

    override fun onFailure(
        webSocket: WebSocket,
        t: Throwable,
        response: Response?,
    ) {
        val attributes = extractAttributes(webSocket)
        WebsocketErrorEvent().emit(logger, attributes)
        delegate.onFailure(webSocket, t, response)
    }

    private fun extractAttributes(socket: WebSocket): Attributes =
        Attributes
            .builder()
            .apply {
                val request = socket.request()
                put(NetworkAttributes.NETWORK_PROTOCOL_NAME, "websocket")
                put(HttpAttributes.HTTP_REQUEST_METHOD, request.method)
                put(UrlAttributes.URL_FULL, request.url.toString())
                put(NetworkAttributes.NETWORK_PEER_PORT, request.url.port.toLong())
            }.build()

    companion object {
        private const val SCOPE = "io.opentelemetry.websocket.events"

        private var logger: Logger =
            OpenTelemetry
                .noop()
                .logsBridge
                .loggerBuilder(SCOPE)
                .build()

        @JvmStatic
        fun configure(openTelemetry: OpenTelemetry) {
            logger =
                openTelemetry.logsBridge
                    .loggerBuilder(SCOPE)
                    .build()
        }
    }
}
